### ISSUE TITLE
🚀 Feature Request: Breakpoint debugging support in `workerd`

### ISSUE THREAD

**Issue by dagnelies** [cloudflare/workerd#371]
Hi,

Basically, I don't know how to debug a worker with VS Code. When clicking "Run/Debug" in VS Code, it basically invokes what is configured in  `.vscode/launch.json`. I just have no idea what should be written there since the dev environment is wrapped by wrangler.

Since most of the people are using VS Code and they probably have a similar problem, I assume it would be a nice addition and improving developer experience significantly. 

I've also posted a related thread here: https://community.cloudflare.com/t/how-to-debug-use-breakpoints-with-vs-code/423775

Thanks for consideration.

**Comment by jasnell** [comment]
Fundamentally everything that is needed for this is in place *except* for the fact that when a breakpoint is hit we have to be able to sleep the request thread. This implies that the dev protocol connection needs to be run on a separate thread. @kentonv probably has the most insight into how to make that work.

**Comment by kentonv** [comment]
I am not sure, but I _think_ we just need to set up a dedicated thread to run the inspector socket listen loop (in server.c++), thinking carefully about synchronization issues. Luckily the types in `worker.h` are already designed to allow multi-threaded access.

**Comment by ohodson** [comment]
> Seems related to this PR, but it has recently been closed without indications as to why: #595

#595 is just enough to get breakpoints working locally, but this also intersects with product code and has some locking issues. It needs a better approach.

**Comment by yacinehmito** [comment]
Thank you very much for the prompt answer.

**Comment by tforster** [comment]
Do we have any update or ETA to enable debugging? My team really want to use Cloudflare Pages and Page Workers but without proper debugging I cannot possibly justify the additional time required to get to market.

**Comment by huw** [comment]
Current PR for this is at #710 (since it wasn’t already linked)

**Comment by huw** [comment]
Here’s the current PR for Wrangler: cloudflare/workers-sdk#3774. It’ll require a Miniflare bump before it can land—but we’re on the home stretch now 🥹

**Comment by abdolrhman** [comment]
I Need this for webstorm, how the old version support attach debugger and this does not, 
using console is big waste of time, thanks for all effort

**Comment by huw** [comment]
We can close this issue, it shipped in Wrangler 3.7.0!

**Comment by ohodson** [comment]
Thanks @huw, that's the last piece.

**Comment by joseph-simpson** [comment]
This is fantastic. Thanks for all the hard work! 

I can set a breakpoint on the JS bundle generated by esbuild (it's something), but of course I really want to set my breakpoint on my TypeScript API (the dream).  

I tried `"sourceMap": true` in my tsconfig.json (as per [stackoverflow](https://stackoverflow.com/questions/76244802/workerd-devtools-enabling-source-maps)) but didn't see any source maps generated.  Also, pages doesn't support worker.toml.  Identifying the right process to attach to is also unclear (I use pnpm to run these in parallel: 1) esbuild src/index.ts --bundle --watch --format=esm --outfile=dist/_worker.js 2) wrangler pages dev dist --live-reload)

After looking through the existing issues, I was going to raise a new issue requesting a short explanation for "**How to breakpoint debugging on TypeScript code running under pages locally**", but I couldn't see a fit with any of the available categories...

Does anyone have any suggestions on how to achieve this or how to raise the requirement?

Thanks again!

**Comment by mrbbot** [comment]
Hey @joseph-simpson! 👋 Try adding `--sourcemap` to your `esbuild` command, and opening the DevTools by pressing <kbd>d</kbd> in your `wrangler pages dev` session. That seemed to work for me on `wrangler@3.7.0`. 🙂

**Comment by tforster** [comment]
@mrbbot What exactly is d supposed to do? I get that it says "open Devtools" but it has never done anything for me. I just assumed it was an uncompleted placeholder for something.

Is there a sample .vscode/launch.json config that can be shared so that we can get debugging working properly in VSCode now?

**Comment by mrbbot** [comment]
Hey @tforster! 👋 <kbd>d</kbd> should open `https://devtools.devprod.cloudflare.dev/js_app?theme=systemPreferred&debugger=true&ws=localhost%3A9229%2Fcore:user:<YOUR_WORKER_NAME>` in a supported browser. I'm guessing you don't have Chrome installed? Firefox should be supported too, but I'm aware of an issue that prevents Wrangler from opening it.

Breakpoint debugging is currently supported using these DevTools, or WebStorm's. I've struggled to get VSCode's debugger working. 😕 VSCode's `attach` configurations connect to `workerd`, but can't find source maps. From reading [VSCode docs](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_source-map-discovery),  it looks like VSCode ignores `//# sourceMappingURL` comments, and instead looks for source maps itself on disk? I may be misunderstanding this though. `wrangler dev` stores its source maps in a random temporary directory that VSCode wouldn't be able to find on its own. I've got some ideas to try work around this though. Will report back on this thread if I get something working. 👍

**Comment by huw** [comment]
The `--no-bundle` flag should work in the meantime, yeah? I’ll give it a try soon and report back if you’re not sure

**Comment by mrbbot** [comment]
Put a PR up to add support for VSCode's debugger to Miniflare: https://github.com/cloudflare/miniflare/pull/681. We'll need to make a few Wrangler changes too, but these should be done soon. 🙂

**Comment by tforster** [comment]
@mrbbot 

> ...I'm guessing you don't have Chrome installed? 

Actually, I do. However, I am running Wrangler in Ubuntu 22.04 under WSL2 on Windows. So Wrangler is on the Ubuntu side but my Chrome is on Windows. I had no idea that "d" was supposed to open Chrome Dev Tools. But now I do I have been able to open it manually. Do you happen to know where the code is that launches Chrome from the "d" keypress (e.g. what repo? Maybe even where in the repo?) If you can point me in that direction I can take a look and perhaps figure out how to get it to work in WSL2 world.

> Breakpoint debugging is currently supported using these DevTools, or WebStorm's. I've struggled to get VSCode's debugger working. 😕 VSCode's `attach` configurations connect to `workerd`, but can't find source maps. From reading [VSCode docs](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_source-map-discovery), it looks like VSCode ignores `//# sourceMappingURL` comments, and instead looks for source maps itself on disk? I may be misunderstanding this though. `wrangler dev` stores its source maps in a random temporary directory that VSCode wouldn't be able to find on its own. I've got some ideas to try work around this though. Will report back on this thread if I get something working. 👍

Breakpoint debugging is something I have taken for granted since I first experienced it with Turbo Pascal on DOS back in the late 80's. So much so that I never stopped to think about how it works. From day 1 JavaScript debugging in VSCode just worked out of the box. And, the Serverless Framework has the serverless-offline plugin making debugging AWS Lambda's a breeze too. CloudFlare workers are the first time I have come across the inability to debug in many years. So, it's really good to understand the intricacies and challenges of the process. I am going to read the VSCode docs you linked so I have a better appreciation and understanding. 

Thanks for all your work on this!!!

**Comment by mrbbot** [comment]
@tforster

> Do you happen to know where the code is that launches Chrome from the "d" keypress?

https://github.com/cloudflare/workers-sdk/blob/a986f19f2d7989639524f9fd73761ea69aef4f6b/packages/wrangler/src/inspect.ts#L853-L900

Thanks for offering to take a look! 😃

**Comment by ben-xD** [comment]
When using webstorm's debugger, the worker freezes and all subsequent requests will time out. The devtools debugger works fine.

**Comment by ohodson** [comment]
@ben-xD can you open a new ticket and post repro instructions for the issue in WebStorm and/or capture a Chrome Debugger Protocol trace in WebStorm? Thanks

**Comment by Phoenixmatrix** [comment]
Just tried wrangler 3.9 which I believe has these changes. Unfortunately it doesn't seem to be quite there yet.

The problem from what I see is that the sourcemap generated for the bundle by wrangler's toolchain doesn't follow along from previous files correctly.  Eg: right now if I run the app with VSCode's debugger, it correctly maps the bundle (under "deployed" in the devtools) to the compiled file my tooling is generating, but if sourcemaps were processed correctly, it should map all the way to the original files:

![image](https://github.com/cloudflare/workerd/assets/1071928/eaa2614b-263f-44e4-8b58-2b16739ccbd3)

You can see in the screenshot that the bundle (ick3smw8et.js) maps correctly to `[[path]].js`. But the later is itself a bundle. This is a remix app, and it has sourcemaps for `//# sourceMappingURL=/build/[[path]].js.map`, and that isn't honored (my knowledge of sourcemap building is limited, but I believe the previous sourcemap in the chain is supposed to be provided when building the next one, so you only need the final one. Maybe that's not happening?)

The issue happens both in Visual Studio Code, or at the Cloudflare devtool url.

If I copy paste the `[[path]].js.map` in my public directory, and in the devtool rightclick in `ick3smw8et.js`, choose add sourcemaps, and add a url to `[[path]].js.map`, then I can debug my source correctly. With the asterisk that its' missing extra lines added during the bundling process, so all my breakpoints are 2-3 lines off.

Considering mapping from my own bundler's output, to wrangler's bundler output, seems to be working correctly, I have a feeling if the final sourcemap was generated using the previous sourcemaps, this would "Just work".

Sorry for the very convoluted explanation! I know just enough about the topic to be dangerous.

**Comment by BrandonNoad** [comment]
@Phoenixmatrix It looks like the `sourceMappingURL` was just fixed by https://github.com/remix-run/remix/pull/7574

With that change, I am able to get the breakpoint debugging working with the Chrome Dev Tools. Though, if you are using Remix, `remix dev -c` swallows the `[b] open a browser, [d] open Devtools, [c] clear console, [x] to exit` menu, so you need to open `https://devtools.devprod.cloudflare.dev/js_app?theme=systemPreferred&ws=localhost%3A9229%2Fws&debugger=true` manually.

I wasn't able to get debugging working inside VS Code though.

**Comment by mrbbot** [comment]
@BrandonNoad, are you on macOS? https://github.com/cloudflare/workers-sdk/pull/4067 will likely need to be merged/released before this works in VSCode 👍

**Comment by BrandonNoad** [comment]
> @BrandonNoad, are you on macOS? https://github.com/cloudflare/workers-sdk/pull/4067 will likely need to be released before this works in VSCode (hopefully going out tomorrow) 👍 

Yes, I'm on a Mac. I'll keep an eye out for the release. Thanks for fixing!

**Comment by mrbbot** [comment]
[`wrangler@3.11.0`](https://github.com/cloudflare/workers-sdk/releases/tag/wrangler%403.11.0) was released yesterday including this fix. I don't think the Remix fix has been released yet, but this should get us one step closer. 👍

**Comment by BrandonNoad** [comment]
> [`wrangler@3.11.0`](https://github.com/cloudflare/workers-sdk/releases/tag/wrangler%403.11.0) was released yesterday including this fix. I don't think the Remix fix has been released yet, but this should get us one step closer. 👍

Awesome. I verified that it works as expected with both patches applied. Thanks for fixing it!

**Comment by JustinGrote** [comment]
My notes for getting vscode working and operating with the vitest vscode test runner extension are here:
https://github.com/cloudflare/workers-sdk/discussions/4174

**Comment by swyxio** [comment]
saving people some time from reading thru a bunch of the above - the vscode + wrangler instructions you need are here. https://blog.cloudflare.com/debugging-cloudflare-workers#breakpoints

**Comment by ReinsBrain** [comment]
I just noticed that I can get DevTools working in windows following instructions at: https://blog.cloudflare.com/debugging-cloudflare-workers#breakpoints ...but not in WSL2 - anybody else have that problem and workaround?

**Comment by erik-beus** [comment]
> I just noticed that I can get DevTools working (but not vscode debugging) in windows following instructions at: https://blog.cloudflare.com/debugging-cloudflare-workers#breakpoints
> 
> ...but DevTools still won't work in WSL2 - anybody else have that problem and workaround? (ps. using Typescript)

I'm not a Windows user, but I also had issues attaching my debugger. For me, downgrading wrangler fixed it - see reported issue here https://github.com/cloudflare/workers-sdk/issues/5297

**Comment by pendevx** [comment]
On one of my projects, I couldn't get either chrome devtools nor vscode to work, no matter what, where on the other, vscode breakpoints were working but devtools wouldn't until vscode's debugger session was started then closed, but even in that case the devtools just showed a bunch of random `eval-somehash.js` files

Honestly could just be some configuration reason with my first project, idk what though, and it's really annoying to keep having to use `console.log` debugging ngl
